### ISSUE TITLE
Incorrect directory permissions for cache

### ISSUE THREAD

**Issue by infertux** [rails/rails#20335]
For some weird reason, Rails seems to be creating directories with incorrect permissions, which doesn't allow to create sub-directories and raises this exception: `Errno::EACCES - Permission denied @ dir_s_mkdir - [...]/tmp/cache/864/4C1`

`tmp/cache/864` is created with permissions `drw-r-xr-x` (655) instead of `drwxr-xr-x` (755) and thus `tmp/cache/864/4C1` cannot be created.

```bash
$ umask
0022
```

```ruby
$ irb
File.umask.to_s(8)
=> "22"
```

```
/usr/local/lib/ruby/2.1.0/fileutils.rb:251:in `mkdir': Permission denied @ dir_s_mkdir - [...]/tmp/cache/864/4C1 (ActionView::Template::Error)
	from /usr/local/lib/ruby/2.1.0/fileutils.rb:251:in `fu_mkdir'
	from /usr/local/lib/ruby/2.1.0/fileutils.rb:224:in `block (2 levels) in mkdir_p'
	from /usr/local/lib/ruby/2.1.0/fileutils.rb:222:in `reverse_each'
	from /usr/local/lib/ruby/2.1.0/fileutils.rb:222:in `block in mkdir_p'
	from /usr/local/lib/ruby/2.1.0/fileutils.rb:208:in `each'
	from /usr/local/lib/ruby/2.1.0/fileutils.rb:208:in `mkdir_p'
	from vendor/bundle/ruby/2.1.0/gems/activesupport-3.2.21/lib/active_support/cache/file_store.rb:162:in `ensure_cache_path'
	from vendor/bundle/ruby/2.1.0/gems/activesupport-3.2.21/lib/active_support/cache/file_store.rb:91:in `write_entry'
	from vendor/bundle/ruby/2.1.0/gems/activesupport-3.2.21/lib/active_support/cache/strategy/local_cache.rb:140:in `write_entry'
	from vendor/bundle/ruby/2.1.0/gems/activesupport-3.2.21/lib/active_support/cache.rb:364:in `block in write'
	from vendor/bundle/ruby/2.1.0/gems/activesupport-3.2.21/lib/active_support/cache.rb:520:in `instrument'
	from vendor/bundle/ruby/2.1.0/gems/activesupport-3.2.21/lib/active_support/cache.rb:362:in `write'
	from vendor/bundle/ruby/2.1.0/gems/actionpack-3.2.21/lib/action_controller/caching/fragments.rb:63:in `block in write_fragment'
	from vendor/bundle/ruby/2.1.0/gems/actionpack-3.2.21/lib/action_controller/caching/fragments.rb:123:in `block in instrument_fragment_cache'
	from vendor/bundle/ruby/2.1.0/gems/activesupport-3.2.21/lib/active_support/notifications.rb:123:in `block in instrument'
	from vendor/bundle/ruby/2.1.0/gems/activesupport-3.2.21/lib/active_support/notifications/instrumenter.rb:20:in `instrument'
	from vendor/bundle/ruby/2.1.0/gems/activesupport-3.2.21/lib/active_support/notifications.rb:123:in `instrument'
	from vendor/bundle/ruby/2.1.0/gems/actionpack-3.2.21/lib/action_controller/caching/fragments.rb:123:in `instrument_fragment_cache'
	from vendor/bundle/ruby/2.1.0/gems/actionpack-3.2.21/lib/action_controller/caching/fragments.rb:61:in `write_fragment'
	from vendor/bundle/ruby/2.1.0/gems/actionpack-3.2.21/lib/action_view/helpers/cache_helper.rb:59:in `fragment_for'
	from vendor/bundle/ruby/2.1.0/gems/actionpack-3.2.21/lib/action_view/helpers/cache_helper.rb:36:in `cache'
[...]
```

```bash
$ bundle exec rails c staging
Loading staging environment (Rails 3.2.21)
irb(main):001:0> Dir.mkdir "tmp/cache/test"
=> 0
irb(main):002:0> puts `ls -l tmp/cache/`
drwxr-xr-x 3 user user 4096 May 28 07:00 2BF
drw-r-xr-x 2 user user 4096 May 28 06:51 864
drwxr-xr-x 2 user user 4096 May 28 07:10 test
```

What's even more weird is that *some* directories (like `tmp/cache/2BF` above) are created with the correct permissions. Any ideas what's going on?

**Comment by infertux** [comment]
Temporary (ugly) fix:

```diff
--- activesupport/lib/active_support/cache/file_store.rb	2014-07-18 15:42:19.427814707 +0000
+++ activesupport/lib/active_support/cache/file_store.rb	2015-05-28 20:24:52.850630507 +0000
@@ -159,7 +159,7 @@
 
         # Make sure a file path's directories exist.
         def ensure_cache_path(path)
-          FileUtils.makedirs(path) unless File.exist?(path)
+          FileUtils.makedirs(path, mode: 0775) unless File.exist?(path)
         end
 
         def search_dir(dir, &callback)
```

Still need to get to the bottom of this...

**Comment by matthewd** [comment]
> Rails 3.2.21

Is this reproducible on a supported version of Rails?

**Comment by infertux** [comment]
@matthewd I don't know but this app will be upgraded to Rails 4 in the next couple of months so we'll see... In the meantime, I'm kinda clueless about the randomness of this bug, I'm not even sure the problem is in Rails...

**Comment by rafaelfranca** [comment]
Can you please provide a sample application that reproduces the error?

**Comment by infertux** [comment]
So I added `raise File.umask.to_s(8) unless File.umask == 022` into `file_store.rb` and it does raise randomly with _umask = 0122_. Since Rails doesn't seem to change `File.umask` anywhere, the issue is probably somewhere else (Unicorn?) so I'm closing this.

[I'll post my findings here...](https://xkcd.com/979/)

**Comment by rubyconvict** [comment]
@matthewd I'am using Rails '4.1.11' with Passenger 4.0.x, on both Ruby 1.9.x and 2.1.x, I also can't remove old release with capistrano because of that. Assets subdirectories in cache dir do have 'w' permission for group, cache subdirectories don't, why?

**Comment by infertux** [comment]
I have yet to get to the bottom of this issue but it seems to be related to how I restart Unicorn. A "zero-downtime restart" (sending `SIGUSR2`) seems to create directories with incorrect permissions whereas a hard restart (`SIGQUIT` then start Unicorn again) seems okay...

**Comment by madej** [comment]
Hi guys. Just to let you know I had the same problem today. It turned out it was monit (mine in version 5.9). If you use monit it might be it.

**Comment by alpracka** [comment]
I've just run in to the same problem. Fresh deploy and I got error:

```
Errno::EACCES (Permission denied @ dir_s_mkdir - /home/app/app/releases/20150909064332/tmp/cache/220/5C0):
  app/models/exchange_rate.rb:115:in `refresh_rates'
  app/controllers/application_controller.rb:21:in `set_conversion_rates'
```

`ls -la` in cache directory

```
drwxr-xr-x 4 caflou caflou 4 Sep  9 08:53 .
drwxr-xr-x 5 caflou caflou 5 Sep  9 08:43 ..
drwxr-xr-x 3 caflou caflou 3 Sep  9 08:48 assets
drw-r-xr-x 2 caflou caflou 2 Sep  9 08:53 220
```

Using **Rails 4.2.4** with Puma 2.13.4.

First line is causing that:

``` ruby
rates = Rails.cache.fetch "rates", expires_in: 12.hours do
  ExchangeRate.update_rates!
  Money.default_bank.save_rates_to_s
end
```

@madej I'm also using Monit, but how it can be related?
@infertux I would reopen this issue.

**Comment by madej** [comment]
Problem is as described above. Monit changes umask of programs it's strarted. Get newer version of monit, they fixed it in 5.10 i believe (i've installed 5.14 and problem is gone).

**Comment by infertux** [comment]
Hmm that's interesting @alpracka - I assumed the problem was due to Unicorn but if you're experiencing it with Puma as well, it must be something else... Pretty sure it has nothing to do with Rails though since if you `grep` through the Rails codebase, it never fiddles with umask. Are you using Monit? I know I was. However, I recently moved the cache to Redis so I won't be able look into this assumption.
